counterexamples/map_floor.lean

```diff
@@ -0,0 +1,126 @@
+* `f` preserves what's between `n` and `n + 1` because it's monotone.
+
+However, that "between" doesn't have to be strict. Potentially something whose floor was `n` could
+get mapped to `n + 1`, and this has floor `n + 1`, not `n`. Note that this is at most an off by one
+error.
+
+This pathology disappears if you require `f` to be strictly monotone or `α` to be archimedean.
+
+## The counterexample
+
+Consider `ℤ[ε]` (`int_with_epsilons`), the integers with infinitesimals adjoined. This is a linearly
+ordered commutative floor ring (`int_with_epsilons.linear_ordered_comm_ring`,
+`int_with_epsilons.floor_ring`).
+
+The map `f : ℤ[ε] → ℤ` that forgets about the epsilons (`int_with_epsilons.forget_epsilons`) is an
+ordered ring homomorphism.
+
+But it does not preserve floors (nor ceils) as `⌊-ε⌋ = -1` while `⌊f (-ε)⌋ = ⌊0⌋ = 0`
+(`int_with_epsilons.forget_epsilons_floor_lt`, `int_with_epsilons.lt_forget_epsilons_ceil`).
+-/
+
+noncomputable theory
+
+open function int polynomial
+open_locale polynomial
+
+/-- The integers with infinitesimals adjoined. -/
+@[derive [comm_ring, nontrivial, inhabited]] def int_with_epsilon := ℤ[X]
+
+notation `ℤ[ε]` := int_with_epsilon
+
+notation `ε` := (X : ℤ[ε])
+
+namespace int_with_epsilon
+
+instance : linear_order ℤ[ε] := linear_order.lift' (to_lex ∘ coeff) coeff_injective
+
+instance : ordered_add_comm_group ℤ[ε] :=
+by refine (to_lex.injective.comp coeff_injective).ordered_add_comm_group _ _ _ _ _ _ _;
+  refl <|> intros; ext; simp [←nsmul_eq_mul, ←zsmul_eq_mul]
+
+lemma pos_iff {p : ℤ[ε]} : 0 < p ↔ 0 < p.trailing_coeff :=
+begin
+  rw trailing_coeff,
+  refine ⟨_, λ h,
+    ⟨p.nat_trailing_degree, λ m hm, (coeff_eq_zero_of_lt_nat_trailing_degree hm).symm, h⟩⟩,
+  rintro ⟨n, hn⟩,
+  convert hn.2,
+  exact (nat_trailing_degree_le_of_ne_zero hn.2.ne').antisymm
+    (le_nat_trailing_degree (by { rintro rfl, cases hn.2.false }) $ λ m hm, (hn.1 _ hm).symm),
+end
+
+instance : linear_ordered_comm_ring ℤ[ε] :=
+{ zero_le_one := or.inr ⟨0, by simp⟩,
+  mul_pos := λ p q, by { simp_rw [pos_iff, trailing_coeff_mul], exact mul_pos },
+  ..int_with_epsilon.linear_order, ..int_with_epsilon.comm_ring,
+  ..int_with_epsilon.ordered_add_comm_group, ..int_with_epsilon.nontrivial }
+
+instance : floor_ring ℤ[ε] :=
+floor_ring.of_floor _ (λ p, if (p.coeff 0 : ℤ[ε]) ≤ p then p.coeff 0 else p.coeff 0 - 1) $ λ p q,
+  begin
+    simp_rw [←not_lt, not_iff_not],
+    split,
+    { split_ifs,
+      { rintro ⟨_ | n, hn⟩,
+        { refine (sub_one_lt _).trans _,
+          simpa using hn },
+        { dsimp at hn,
+          simp [hn.1 _ n.zero_lt_succ] } },
+      { exact λ h', cast_lt.1 ((not_lt.1 h).trans_lt h') } },
+    { split_ifs,
+      { exact λ h', h.trans_le (cast_le.2 $ sub_one_lt_iff.1 h') },
+      { exact λ h', ⟨0, by simpa using h'⟩ } }
+  end
+
+/-- The ordered ring homomorphisms from `ℤ[ε]` to `ℤ` that "forgets" the `ε`s. -/
+def forget_epsilons : ℤ[ε] →+*o ℤ :=
```

Comment:
Too bad that https://leanprover-community.github.io/mathlib_docs/data/mv_polynomial/basic.html#mv_polynomial.constant_coeff doesn't seem to exist for `polynomial`.

Reply:
Do you want me to write it?

Reply:
Let's do it in a future PR